pandas/core/reshape/merge.py

```diff
@@ -2377,7 +2379,10 @@ def _factorize_keys(
         rk = ensure_int64(rk.codes)
 
     elif isinstance(lk, ExtensionArray) and is_dtype_equal(lk.dtype, rk.dtype):
-        if not isinstance(lk, BaseMaskedArray):
+        if not isinstance(lk, BaseMaskedArray) and not (
+            isinstance(lk.dtype, ArrowDtype)
+            and is_numeric_dtype(lk.dtype.numpy_dtype.type)
```

Comment:
```suggestion
            and lk.dtype._is_numeric
```
Might simpler?

Reply:
doesn't work unfortunately, we have to exclude all arrow dtypes whose numpy dtype is object